lib/parsers/ogg_parser.rb

```diff
@@ -0,0 +1,190 @@
+
+  # The maximum size of an Ogg page is 65,307 bytes.
+  MAX_POSSIBLE_PAGE_SIZE = 65307
+
+  def call(io)
+    # The format consists of chunks of data each called an "Ogg page". Each page
+    # begins with the characters, "OggS", to identify the file as Ogg format.
+    capture_pattern = safe_read(io, 4)
+    return unless capture_pattern == 'OggS'
+
+    io.seek(28) # skip not important bytes
+
+    # Each header packet begins with the same header fields.
+    #   1) packet_type: 8 bit value (the identification header is type 1)
+    #   2) the characters v','o','r','b','i','s' as six octets
+    packet_type, vorbis, _vorbis_version, channels, sample_rate = safe_read(io, 16).unpack('Ca6LCL')
+    return unless packet_type == 1 && vorbis == 'vorbis'
+
+    # In order to calculate the audio duration we have to read a
+    # granule_position of the last Ogg page of the file. Unfortunately, we don't
+    # know where the last page starts. But we do know that max size of an Ogg
+    # page is 65307 bytes. So we read the last 65307 bytes from the file and try
+    # to find the last page in this tail.
+    pos = io.size - MAX_POSSIBLE_PAGE_SIZE
+    pos = 0 if pos < 0
+    io.seek(pos)
+    tail = io.read(MAX_POSSIBLE_PAGE_SIZE)
+    granule_position = find_granule_position(tail)
+    return if granule_position.nil?
+
+    duration = granule_position / sample_rate.to_f
+    return if duration == Float::INFINITY
+
+    FormatParser::Audio.new(
+      format: :ogg,
+      audio_sample_rate_hz: sample_rate,
+      num_audio_channels: channels,
+      media_duration_seconds: duration
+    )
+  end
+
+  private
+
+  # Returns granule_position of the last valid Ogg page contained in the given
+  # tail. Since the tail may contain multiple "OggS" entries the method searches
+  # them recursively starting from the end. The search stops when the first
+  # valid Oggs page is found.
+  def find_granule_position(tail)
+    # The Ogg page always starts with "OggS".
+    pos_of_the_last_page = tail.rindex('OggS')
+
+    # Return if for some reason the tail doesn't contain the magic bits.
+    return if pos_of_the_last_page.nil?
+
+    # Since the magic bits may occur inside the body of the page we have to
+    # validate that what we found is actually an Ogg page by calculating the
+    # checksum. For this reason we have to read the entire page and calculate
+    # its checksum. In order to read the entire Ogg page we first have to read a
+    # part of its header to find out the size of the page.
+
+    # Read the Ogg page header excluding the segment table (in other words read
+    # first 27 bytes).
+    header = tail[pos_of_the_last_page...pos_of_the_last_page + 27]
+    return find_granule_position(tail[0...pos_of_the_last_page]) if header.size < 27
+
+    _capture_pattern,
+    _version,
+    _header_type,
+    granule_position,
+    _bitstream_serial_number,
+    _page_sequence_number,
+    checksum,
+    page_segments = header.unpack('a4CCQVVVC')
```

Comment:
_I am copying [my previous reply](https://github.com/WeTransfer/format_parser/pull/116#discussion_r191242229) because it was lost in outdated comments._

@julik this is the last part if I didn't miss anything, but I'm stuck on it. I think it should be fine to use the native endianness, because the docs don't specify it, but I don't know how to test it on a big-endian system so not 100% sure.

The is a place in the docs that specifies endianness:

> However, since data in the comment header is octetaligned,
they can simply be read as unaligned 32 bit little endian unsigned integers.

But it is about the Vorbis text comment header which I don't use.

If you could help me with this part it would be great!